### PR TITLE
[FEATURE] Support Flink RSS.

### PR DESCRIPTION
**Pull request by slfan1989**
### What changes were proposed in this pull request?

# I. Introduction

## A. Background

1. Limitations or issues with the current system
2. Significance and necessity of the proposed solution

image

- IntermediateDataSet

An intermediate data set is the data set produced by an operator - either a source or any intermediate operation.

### Why are the changes needed?

(Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, describe the bug.)

Fix: # (issue)

### Does this PR introduce _any_ user-facing change?

(Please list the user-facing changes introduced by your change, including
  1. Change in user-facing APIs.
  2. Addition or removal of property keys.)

No.

### How was this patch tested?

(Please test your changes, and provide instructions on how to test it:
  1. If you add a feature or fix a bug, add a test to cover your changes. 
  2. If you fix a flaky test, repeat it for many times to prove it works.)

### CODE REVIEW

**Comment by slfan1989** [comment]
@jerqi @zuston I started to submit the code about Flink rss. I still need to spend some time merging the local code. It will take about 2-3 days to complete the submission.

**Comment by slfan1989** [comment]
@chenglong-bigdata Can you help review the code?

**Comment by slfan1989** [comment]
buffer: https://flink.apache.org/2019/06/05/a-deep-dive-into-flinks-network-stack/

File: client-flink/common/src/main/java/org/apache/uniffle/shuffle/RssShuffleMaster.java [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+   * job configuration. It enables ShuffleMaster to notify JobMaster about lost result partitions,
+   * so that JobMaster can identify and reproduce unavailable partitions earlier.
+   *
+   * @param context the corresponding shuffle context of the target job.
+   */
+  @Override
+  public void registerJob(JobShuffleContext context) {
+    JobID jobID = context.getJobId();
+    if (shuffleWriteClient == null) {
+      synchronized (RssShuffleMaster.class) {
```
[the remark this answers is not in the document]
Comment (slfan1989) [reply client-flink/common/src/main/java/org/apache/uniffle/shuffle/RssShuffleMaster.java:124]: Thank you helping to review the code! This PR still needs to be improved.

**Comment by chenglong-bigdata** [comment]
> @chenglong-bigdata Can you help review the code?

ok

**Comment by slfan1989** [comment]
> Is there a design doc for this one? I can help review this feature.

Thank you for reviewing the code! The documentation is being improved. I will update it as soon as possible. Unit tests are still being added.

**Comment by slfan1989** [comment]
I will continue to follow up on this PR.

**Comment by slfan1989** [comment]
@chenglong-bigdata We will continue to follow up on this PR.

**Pull request closed without merging**